File: python-importer/deepbrain.py
"""深脑上传客户端：走现成的录音链路，不需要给深脑加任何接口。

认证用「原生端 Bearer」通道（lib/recording/route-context.ts 明确支持）：
    Authorization: Bearer <supabase access token>
    x-deepbrain-org-id: <org uuid>

上传五步（对应深脑 apps/web/src/app/api/recordings/*）：
    1. POST /api/recordings                              建会话
    2. POST /api/recordings/{id}/chunks/ticket           要一个分片上传地址
    3. PUT  <uploadUrl>                                  直传腾讯云 COS，不过深脑服务器
    4. POST /api/recordings/{id}/chunks/{seq}/complete   服务端校验分片
    5. POST /api/recordings/{id}/stop                    冻结分片总数与时长（不做会 409）
    6. POST /api/recordings/{id}/finalize                收尾，之后自动进说话人分离与分析

密码只在系统对话框里输入，直接换成 token，本工具不落盘、不打印密码。
refresh token 存 macOS 钥匙串（服务名 deepbrain-importer）。
"""
from __future__ import annotations

import json
import subprocess
import urllib.error
import urllib.request
import uuid
from dataclasses import dataclass
from pathlib import Path

CONFIG = json.loads((Path(__file__).resolve().parent / "deepbrain.json").read_text())
API = CONFIG["apiBase"].rstrip("/")
SB = CONFIG["supabaseUrl"].rstrip("/")
ANON = CONFIG["supabaseAnonKey"]

# 凭证改存文件，不再用钥匙串。
# 原因：钥匙串授权绑代码签名身份，而 ad-hoc 签名每次重新部署身份都变，
# macOS 每次都当陌生程序弹框问登录密码，点「始终允许」也没用。
# 代价：文件 0600，只有本用户可读，比钥匙串弱；存的是 refresh token 不是密码，
# 泄露可在深脑吊销。Swift 版读的是同一个文件，两边不会各存一份。
CRED_FILE = Path.home() / "Library/Application Support/深脑/credentials.json"
_KEY_MAP = {"refresh_token": "refreshToken", "email": "email", "org_id": "orgId"}
MAX_CHUNK_BYTES = 8 * 1024 * 1024      # 与深脑 service.ts 的 MAX_CHUNK_BYTES 对齐


class DeepBrainError(RuntimeError):
    pass


# --- HTTP ---
def _request(method: str, url: str, *, headers: dict | None = None,
             json_body: dict | None = None, data: bytes | None = None,
             timeout: float = 120) -> tuple[int, bytes]:
    body = data
    hdr = dict(headers or {})
    if json_body is not None:
        body = json.dumps(json_body).encode()
        hdr["Content-Type"] = "application/json"
    req = urllib.request.Request(url, data=body, method=method, headers=hdr)
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            return resp.status, resp.read()
    except urllib.error.HTTPError as exc:
        return exc.code, exc.read()


def _json_or_raise(status: int, raw: bytes, what: str) -> dict:
    if status >= 400:
        detail = raw.decode("utf-8", "replace")[:300]
        raise DeepBrainError(f"{what} 失败 HTTP {status}：{detail}")
    return json.loads(raw or b"{}")


# --- 钥匙串 ---
def _cred_read() -> dict:
    try:
        return json.loads(CRED_FILE.read_text())
    except Exception:
        return {}


def _keychain_get(account: str = "refresh_token") -> str | None:
    return _cred_read().get(_KEY_MAP.get(account, account)) or None


def _keychain_set(value: str, account: str = "refresh_token") -> None:
    blob = _cred_read()
    blob[_KEY_MAP.get(account, account)] = value
    CRED_FILE.parent.mkdir(parents=True, exist_ok=True)
    CRED_FILE.write_text(json.dumps(blob, ensure_ascii=False))
    CRED_FILE.chmod(0o600)


def _prompt(label: str, hidden: bool = False) -> str:
    """弹到最前面，否则对话框会藏在别的窗口后面，看起来像卡住了。"""
    hidden_clause = " with hidden answer" if hidden else ""
    script = (
        'tell application "System Events"\n'
        '  activate\n'
        f'  set r to display dialog "{label}" default answer "" '
        f'with title "深脑登录" with icon note{hidden_clause}\n'
        '  return text returned of r\n'
        'end tell'
    )
    r = subprocess.run(["osascript", "-e", script], capture_output=True, text=True)
    if r.returncode != 0:
        raise DeepBrainError("已取消登录" if "-128" in r.stderr else f"弹窗失败：{r.stderr.strip()[:120]}")
    return r.stdout.strip()


# --- 认证 ---
def login() -> str:
    """弹系统对话框让用户自己输入。密码只用于换 token，不落盘。"""
    email = _prompt("深脑账号（邮箱）")
    password = _prompt("深脑密码", hidden=True)
    status, raw = _request("POST", f"{SB}/auth/v1/token?grant_type=password",
                           headers={"apikey": ANON},
                           json_body={"email": email, "password": password})
    data = _json_or_raise(status, raw, "登录")
    if not data.get("refresh_token"):
        raise DeepBrainError("登录应答里没有 refresh_token")
    _keychain_set(data["refresh_token"])
    _keychain_set(email, "email")
    return data["access_token"]


def access_token() -> str:
    """用钥匙串里的 refresh token 换 access token，并轮换保存。"""
    refresh = _keychain_get()
    if not refresh:
        raise DeepBrainError("尚未登录。先运行：python deepbrain.py login")
    status, raw = _request("POST", f"{SB}/auth/v1/token?grant_type=refresh_token",
                           headers={"apikey": ANON}, json_body={"refresh_token": refresh})
    if status >= 400:
        raise DeepBrainError("登录已失效，请重新运行：python deepbrain.py login")
    data = json.loads(raw)
    if data.get("refresh_token"):
        _keychain_set(data["refresh_token"])
    return data["access_token"]


def org_id(token: str) -> str:
    """从 memberships 表取所属组织（深脑原生端要求显式指定 org）。"""
    cached = _keychain_get("org_id")
    if cached:
        return cached
    status, raw = _request("GET", f"{SB}/rest/v1/memberships?select=org_id&limit=1",
                           headers={"apikey": ANON, "Authorization": f"Bearer {token}"})
    rows = _json_or_raise(status, raw, "查询组织")
    if not rows:
        raise DeepBrainError("这个账号没有任何组织，先在深脑网页里初始化")
    oid = rows[0]["org_id"]
    _keychain_set(oid, "org_id")
    return oid


@dataclass
class Session:
    token: str
    org: str

    @property
    def headers(self) -> dict:
        return {"Authorization": f"Bearer {self.token}", "x-deepbrain-org-id": self.org}


def connect() -> Session:
    token = access_token()
    return Session(token=token, org=org_id(token))


# --- 上传 ---
def upload_recording(sess: Session, path: Path, title: str, duration_sec: float,
                     mime: str = "audio/ogg", on_step=None) -> dict:
    """把一个已录好的音频文件当作一次录音会话推给深脑。

    1 小时 opus 约 7.2MB，在单片 8MB 上限内，通常只需要一个分片。
    """
    audio = path.read_bytes()
    if not audio:
        raise DeepBrainError(f"{path.name} 是空文件")
    step = on_step or (lambda _m: None)

    step("建会话")
    status, raw = _request("POST", f"{API}/api/recordings", headers=sess.headers, json_body={
        "clientRequestId": f"mac-ble-{path.stem}",   # 幂等键：同一文件重传不会建两个会话
        "title": title,
        "captureClient": "macos",
        "capabilities": ["mic"],
    })
    session = _json_or_raise(status, raw, "建会话")["session"]
    session_id = session["id"]

    # clientRequestId 是幂等键：同一文件重推会拿回同一个会话。
    # 若它早已 stop/finalize，就不能再塞分片（服务端 409 INVALID_STATE）——
    # 这不是错误，是「已经做完了」。直接返回，别把重试变成失败。
    if session.get("status") not in ("recording", "uploading"):
        step(f"会话已是 {session.get('status')}，无需重传")
        return {"sessionId": session_id, "chunks": session.get("expected_chunk_count") or 0,
                "bytes": len(audio), "alreadyDone": True,
                "url": f"{API}/recordings/{session_id}"}

    total_ms = max(1, int(duration_sec * 1000))
    parts = [audio[i:i + MAX_CHUNK_BYTES] for i in range(0, len(audio), MAX_CHUNK_BYTES)]
    for seq, part in enumerate(parts):
        step(f"分片 {seq + 1}/{len(parts)}")
        started = total_ms * seq // len(parts)
        ended = total_ms * (seq + 1) // len(parts)
        status, raw = _request("POST", f"{API}/api/recordings/{session_id}/chunks/ticket",
                               headers=sess.headers, json_body={
                                   "sequence": seq,
                                   "idempotencyKey": f"{path.stem}-{seq}-{uuid.uuid5(uuid.NAMESPACE_URL, path.name).hex[:8]}",
                                   "mimeType": mime,
                                   "byteLength": len(part),
                                   "startedAtMs": started,
                                   "endedAtMs": max(ended, started + 1),
                                   "uploadMode": "background",
                               })
        if status == 409 and b"CHUNK_ALREADY_VERIFIED" in raw:
            step(f"分片 {seq + 1} 已在服务端，跳过")   # 重试时的正常情况，不是错误
            continue
        ticket = _json_or_raise(status, raw, "申请上传地址")

        up_status, up_raw = _request("PUT", ticket["uploadUrl"], data=part,
                                     headers={"Content-Type": mime}, timeout=600)
        if up_status >= 400:
            raise DeepBrainError(f"直传 COS 失败 HTTP {up_status}："
                                 f"{up_raw.decode('utf-8','replace')[:200]}")

        status, raw = _request("POST",
                               f"{API}/api/recordings/{session_id}/chunks/{seq}/complete",
                               headers=sess.headers, json_body={})
        _json_or_raise(status, raw, f"确认分片 {seq}")

    # 必须先 stop 冻结「分片总数 + 时长」，否则 finalize 会 409「录音尚未冻结分片数量」。
    # 这是深脑录音链路的核心不变量：清单一旦冻结，后到的分片就再也进不来。
    step("冻结清单")
    status, raw = _request("POST", f"{API}/api/recordings/{session_id}/stop",
                           headers=sess.headers, json_body={
                               "durationMs": total_ms,
                               "expectedChunkCount": len(parts),
                           })
    _json_or_raise(status, raw, "冻结清单")

    step("收尾")
    status, raw = _request("POST", f"{API}/api/recordings/{session_id}/finalize",
                           headers=sess.headers, json_body={})
    _json_or_raise(status, raw, "收尾")
    return {"sessionId": session_id, "chunks": len(parts), "bytes": len(audio),
            "alreadyDone": False, "url": f"{API}/recordings/{session_id}"}


if __name__ == "__main__":
    import sys
    cmd = sys.argv[1] if len(sys.argv) > 1 else "whoami"
    if cmd == "login":
        login()
        print("登录成功，凭证已存入 macOS 钥匙串")
    elif cmd == "whoami":
        s = connect()
        print(f"账号 {_keychain_get('email')}\n组织 {s.org}\n深脑 {API}")
    else:
        print(__doc__)

File: Sources/LuyinbiCore/DeepBrain.swift
import Foundation
import Security

/// 深脑客户端。走现成的录音链路，不需要给深脑加任何接口。
///
/// 认证用「原生端 Bearer」通道（深脑 lib/recording/route-context.ts 明确支持）：
///   Authorization: Bearer <supabase access token> + x-deepbrain-org-id
///
/// 上传六步：建会话 → 要分片地址 → 直传 COS → 确认分片 → **stop 冻结清单** → finalize。
/// 漏掉 stop，finalize 必然 409「录音尚未冻结分片数量」。
public struct DeepBrainConfig: Codable {
    public let apiBase: String
    public let supabaseUrl: String
    public let supabaseAnonKey: String

    public static func load(from url: URL) throws -> DeepBrainConfig {
        try JSONDecoder().decode(DeepBrainConfig.self, from: Data(contentsOf: url))
    }
}

public enum DeepBrainError: Error, CustomStringConvertible {
    case notLoggedIn
    case http(String, Int, String)
    case badResponse(String)
    /// 录音链路来的转写是「权威数据」，深脑只允许服务端改。
    case canonicalReadOnly
    /// 这条已经分析过：改名要连原子主语、决策归属一起重挂，只能走网页。
    case alreadyAnalyzed

    public var description: String {
        switch self {
        case .notLoggedIn: return "尚未登录深脑"
        case .http(let what, let code, let body): return "\(what) 失败 HTTP \(code)：\(body.prefix(200))"
        case .badResponse(let s): return "应答不符合预期：\(s)"
        case .canonicalReadOnly:
            return "这条录音的说话人只能在深脑网页里指认——录音链路的数据是权威数据，客户端改不了。"
        case .alreadyAnalyzed:
            return "这条已经分析过了，改名要连洞察里的归属一起重挂，得在深脑网页里改。"
        }
    }
}

/// 钥匙串。与 Python 版共用同一个服务名，登录状态互通。
public enum Keychain {
    public static let service = "deepbrain-importer"

    public static func get(_ account: String = "refresh_token") -> String? {
        let q: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var item: CFTypeRef?
        guard SecItemCopyMatching(q as CFDictionary, &item) == errSecSuccess,
              let data = item as? Data else { return nil }
        return String(data: data, encoding: .utf8)
    }

    @discardableResult
    public static func set(_ value: String, account: String = "refresh_token") -> Bool {
        let base: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
        ]
        SecItemDelete(base as CFDictionary)
        var add = base
        add[kSecValueData as String] = Data(value.utf8)
        return SecItemAdd(add as CFDictionary, nil) == errSecSuccess
    }
}

public final class DeepBrain {
    // 模块内可见：Speakers.swift 里的扩展要用（Swift 的 private 是文件级的）
    let config: DeepBrainConfig
    var token: String?
    var org: String?
    /// transcript_id -> recording_session_id。指认接口按会话寻址，而 App 全程拿的是
    /// transcript_id；这个映射一条录音只查一次。
    var sessionIdCache: [String: String] = [:]

    public init(config: DeepBrainConfig) { self.config = config }

    /// 真实录音时刻用 ISO8601 传给深脑。
    static let iso: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    /// 专用会话：**绕开系统代理**。
    ///
    /// 深脑跑在国内服务器上（122.51.221.171）。这台 Mac 常年开着翻墙代理
    /// （HTTP/HTTPS/SOCKS 全指 127.0.0.1:7897），而 URLSession.shared 默认跟随系统代理——
    /// 于是本该 147ms 直达的请求被绕进代理，21 MB 的分片上传中途就断，
    /// 日志里只留下 "The network connection was lost."，看起来像深脑挂了。
    ///
    /// 实测：直连 147ms / 200，走代理反复失败。国内的自家服务器没有任何理由走代理。
    /// `connectionProxyDictionary = [:]` 是显式清空，不是"用默认"——默认就是跟随系统。
    ///
    /// 超时给得比默认长：录音分片单个可以到几 MB，弱网下 60s 的默认请求超时不够。
    static let session: URLSession = {
        let c = URLSessionConfiguration.default
        c.connectionProxyDictionary = [:]
        c.timeoutIntervalForRequest = 120
        c.timeoutIntervalForResource = 3600      // 三小时录音整体上传的上限
        c.waitsForConnectivity = true
        return URLSession(configuration: c)
    }()

    // MARK: - HTTP
    func request(_ method: String, _ url: String, headers: [String: String] = [:],
                         json: Any? = nil, body: Data? = nil,
                         timeout: TimeInterval = 300) async throws -> (Int, Data) {
        var req = URLRequest(url: URL(string: url)!, timeoutInterval: timeout)
        req.httpMethod = method
        for (k, v) in headers { req.setValue(v, forHTTPHeaderField: k) }
        if let json {
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
            req.httpBody = try JSONSerialization.data(withJSONObject: json)
        } else if let body {
            req.httpBody = body
        }
        let (data, resp) = try await Self.session.data(for: req)
        return ((resp as? HTTPURLResponse)?.statusCode ?? 0, data)
    }

    private func json(_ status: Int, _ data: Data, _ what: String) throws -> Any {
        guard status < 400 else {
            throw DeepBrainError.http(what, status, String(decoding: data, as: UTF8.self))
        }
        return try JSONSerialization.jsonObject(with: data.isEmpty ? Data("{}".utf8) : data)
    }

    // MARK: - 认证
    public func connect() async throws {
        guard let refresh = TokenStore.get() else { throw DeepBrainError.notLoggedIn }
        let (st, data) = try await request(
            "POST", "\(config.supabaseUrl)/auth/v1/token?grant_type=refresh_token",
            headers: ["apikey": config.supabaseAnonKey], json: ["refresh_token": refresh])
        guard st < 400, let d = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let access = d["access_token"] as? String else {
            throw DeepBrainError.http("刷新登录", st, "登录已失效，请用 Python 版重新登录")
        }
        if let newRefresh = d["refresh_token"] as? String { TokenStore.set(newRefresh) }
        token = access
        org = try await resolveOrg(access)
    }

    func resolveOrg(_ access: String) async throws -> String {
        if let cached = TokenStore.get("org_id") { return cached }
        // 2026-09-02 事故（先在安卓端撞上，这里是同一处漏洞的另一份实现）：
        // 查询原来没有 order by，账号如果有多条 membership，PostgREST
        // 不保证返回顺序——选到哪个组织全凭运气，选错了不会报任何错，
        // 只是那之后每一次查询都在查一个没有数据的空组织。
        // 这里比安卓端更危险：选中的结果会被 TokenStore 缓存写进
        // credentials.json，选错一次就永久错下去，直到手动 TokenStore.clear()。
        // 加 order=created_at.asc：最早那条 membership 几乎总是账号真正在用的那个。
        let (st, data) = try await request(
            "GET", "\(config.supabaseUrl)/rest/v1/memberships?select=org_id&order=created_at.asc&limit=1",
            headers: ["apikey": config.supabaseAnonKey, "Authorization": "Bearer \(access)"])
        guard let rows = try json(st, data, "查询组织") as? [[String: Any]],
              let oid = rows.first?["org_id"] as? String else {
            throw DeepBrainError.badResponse("这个账号没有任何组织")
        }
        TokenStore.set(oid, "org_id")
        return oid
    }

    /// 直连 PostgREST 用的头（走用户登录态 + RLS），与走深脑 API 的 authHeaders 不同。
    var restHeaders: [String: String] {
        get throws {
            guard let token else { throw DeepBrainError.notLoggedIn }
            return ["apikey": config.supabaseAnonKey, "Authorization": "Bearer \(token)"]
        }
    }

    var configSupabaseURL: String { config.supabaseUrl.hasSuffix("/")
        ? String(config.supabaseUrl.dropLast()) : config.supabaseUrl }

    // 模块内可见：Speakers.swift 里的指认接口也走深脑 API 通道
    /// 带自动续期的请求。**所有需要登录的调用都该走它。**
    ///
    /// Supabase 的 access token 只活一小时，而 `connect()` 只在启动时刷新一次，
    /// 之后一直用同一个。于是开着 App 超过一小时之后：
    ///   · 直连 Supabase 的查询回 401 `JWT expired`（界面上是红字报错）
    ///   · 推送用的是同一个 token → **「推送卡住」**
    /// 两个现象同一个根因。2026-09-01 用户截图里就是这一条。
    ///
    /// 只续一次：续了还是 401 说明真的登录失效了，反复续只会空转，
    /// 而正确的做法是让用户重新登录。
    func authed(_ method: String, _ url: String,
                headers: () throws -> [String: String],
                json: Any? = nil, body: Data? = nil,
                timeout: TimeInterval = 300) async throws -> (Int, Data) {
        var (st, data) = try await request(method, url, headers: try headers(),
                                           json: json, body: body, timeout: timeout)
        if st == 401 {
            try await connect()
            // headers 是个闭包，重算一次才能拿到刚续出来的 token。
            // 传值进来的话，重试用的还是那个已经过期的。
            (st, data) = try await request(method, url, headers: try headers(),
                                           json: json, body: body, timeout: timeout)
        }
        return (st, data)
    }

    var authHeaders: [String: String] {
        get throws {
            guard let token, let org else { throw DeepBrainError.notLoggedIn }
            return ["Authorization": "Bearer \(token)", "x-deepbrain-org-id": org]
        }
    }

    // MARK: - 上传
    public struct UploadResult {
        public let sessionId: String
        public let chunks: Int
        /// 会话此前已完成，本次是幂等重放，什么都没重传。
        public let alreadyDone: Bool
    }

    /// 单片上限 8MB（与深脑 service.ts 的 MAX_CHUNK_BYTES 对齐）。
    /// 1 小时录音的 opus 约 7.2MB，通常一片装得下。
    public static let maxChunkBytes = 8 * 1024 * 1024

    /// 上传一条录音。
    ///
    /// `startedAt` 是**录音发生的时刻**，不是上传时刻。不传的话深脑那边
    /// `recording_sessions.started_at` 走的是 `default now()`——于是一条昨天下午的会
    /// 在深脑里显示成今天中午，因为那才是它被推上去的时间。真实时刻就写在文件名里
    /// （note20260829-170356），从来只是没传过去。
    public func upload(audio: [UInt8], title: String, durationSec: Double,
                       clientRequestId: String, mime: String = "audio/ogg",
                       startedAt: Date? = nil,
                       projectId: String? = nil,
                       onStep: ((String) -> Void)? = nil) async throws -> UploadResult {
        onStep?("建会话")
        let (st1, d1) = try await authed("POST", "\(config.apiBase)/api/recordings", headers: { try self.authHeaders }, json: [
            "clientRequestId": clientRequestId,      // 幂等键：同一文件重传不会建两个会话
            "title": title, "captureClient": "macos", "capabilities": ["mic"],
            // 服务端目前还不收这个字段，多传无害；等接口加上就自动生效。
            // 先从客户端把真相带上，比等两边同时改要稳。
        ].merging(startedAt.map { ["startedAt": Self.iso.string(from: $0)] } ?? [:]) { a, _ in a }
         .merging(projectId.map { ["projectId": $0] } ?? [:]) { a, _ in a })
        guard let o1 = try json(st1, d1, "建会话") as? [String: Any],
              let session = o1["session"] as? [String: Any],
              let sid = session["id"] as? String else { throw DeepBrainError.badResponse("建会话没返回 id") }

        // clientRequestId 是幂等键：同一文件重推会拿回同一个会话。
        // 如果它早已 stop/finalize，就不能再塞分片了（服务端会 409 INVALID_STATE）——
        // 这不是错误，是「已经做完了」。直接返回，别把重试变成失败。
        let existing = session["status"] as? String ?? ""
        // failed 不是「已经做完了」。以前这里把它一并当成完成，于是一条失败的会话
        // 会永远挡住重传——客户端每次都拿回同一个 failed 会话，报「无需重传」，
        // 而那条录音其实一个字都没进深脑。失败必须抛出来让人看见。
        if existing == "failed" {
            throw DeepBrainError.badResponse(
                "深脑那边这条会话是 failed 状态（\(sid)）。它挡着重传——"
                + "换一个幂等键重推，或先在深脑里删掉这条会话。")
        }
        if !["recording", "uploading"].contains(existing) {
            onStep?("会话已是 \(existing)，无需重传")
            return UploadResult(sessionId: sid, chunks: session["expected_chunk_count"] as? Int ?? 0,
                                alreadyDone: true)
        }

        let totalMs = max(1, Int(durationSec * 1000))
        let parts = stride(from: 0, to: audio.count, by: Self.maxChunkBytes).map {
            Array(audio[$0..<min($0 + Self.maxChunkBytes, audio.count)])
        }
        for (seq, part) in parts.enumerated() {
            onStep?("分片 \(seq + 1)/\(parts.count)")
            let started = totalMs * seq / parts.count
            let ended = max(totalMs * (seq + 1) / parts.count, started + 1)
            let (st2, d2) = try await authed(
                "POST", "\(config.apiBase)/api/recordings/\(sid)/chunks/ticket",
                headers: { try self.authHeaders }, json: [
                    "sequence": seq, "idempotencyKey": "\(clientRequestId)-\(seq)",
                    "mimeType": mime, "byteLength": part.count,
                    "startedAtMs": started, "endedAtMs": ended, "uploadMode": "background",
                ])
            if st2 == 409, String(decoding: d2, as: UTF8.self).contains("CHUNK_ALREADY_VERIFIED") {
                onStep?("分片 \(seq + 1) 已在服务端，跳过")     // 重试时的正常情况
                continue
            }
            guard let ticket = try json(st2, d2, "申请上传地址") as? [String: Any],
                  let uploadUrl = ticket["uploadUrl"] as? String else {
                throw DeepBrainError.badResponse("ticket 没有 uploadUrl")
            }
            let (st3, d3) = try await request("PUT", uploadUrl,
                                              headers: ["Content-Type": mime], body: Data(part), timeout: 600)
            guard st3 < 400 else {
                throw DeepBrainError.http("直传 COS", st3, String(decoding: d3, as: UTF8.self))
            }
            let (st4, d4) = try await authed(
                "POST", "\(config.apiBase)/api/recordings/\(sid)/chunks/\(seq)/complete",
                headers: { try self.authHeaders }, json: [:])
            _ = try json(st4, d4, "确认分片 \(seq)")
        }

        // 必须先 stop 冻结「分片总数 + 时长」，否则 finalize 会 409。
        // 这是深脑录音链路的核心不变量：清单一旦冻结，后到的分片再也进不来。
        onStep?("冻结清单")
        let (st5, d5) = try await authed("POST", "\(config.apiBase)/api/recordings/\(sid)/stop",
                                         headers: { try self.authHeaders },
                                         json: ["durationMs": totalMs, "expectedChunkCount": parts.count])
        _ = try json(st5, d5, "冻结清单")

        onStep?("收尾")
        let (st6, d6) = try await authed("POST", "\(config.apiBase)/api/recordings/\(sid)/finalize",
                                         headers: { try self.authHeaders }, json: [:])
        _ = try json(st6, d6, "收尾")
        return UploadResult(sessionId: sid, chunks: parts.count, alreadyDone: false)
    }

    public struct SessionState {
        public let status: String
        public let transcriptId: String?
        public let errorCode: String?
    }

    /// CB08 等自带硬件的一步式绑定（spec 019）。deviceNo 是用户现起的名字，
    /// 不是设备自己报的号——CB08 没有真实序列号。
    ///
    /// `.deviceTaken` 单独区分出来：那不是网络失败，是"这个名字已经被
    /// 另一个账号绑过了"，调用方要能就地换个名字重试，而不是报一句
    /// 笼统的"失败了"。
    public enum DeviceBindOutcome {
        case ok
        case deviceTaken
        case failed(String)
    }

    public func selfRegisterDevice(provider: String, deviceNo: String) async -> DeviceBindOutcome {
        do {
            let (st, data) = try await authed(
                "POST", "\(config.apiBase)/api/devices/self-register",
                headers: { try self.authHeaders },
                json: ["provider": provider, "deviceNo": deviceNo])
            if st < 400 { return .ok }
            let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
            let code = obj?["error"] as? String
            switch code {
            case "device_taken": return .deviceTaken
            case "device_no_invalid": return .failed("这个名字里有空格或特殊字符，换一个")
            default: return .failed("绑定失败（HTTP \(st)）")
            }
        } catch {
            return .failed(error.localizedDescription)
        }
    }

    public func sessionState(_ id: String) async throws -> SessionState {
        // 2026-09-02 事故：这里原来走的是 request()（只用当时那份 token，
        // 401 也不会重试），而它是 Monitor 后台轮询循环里的调用点——
        // 会话开着超过一小时，access token 过期之后，同一批会话 id
        // 会以 46 秒一轮的节奏被永久 401 下去，而 refreshBrainStatus 的
        // `try?` 把每一次失败都悄悄吞掉，界面上什么都不会显示，
        // 用户唯一能看到的是"分析状态一直不更新"，看不出原因。
        // upload() 那批调用今天早些时候已经改成 authed()，这条路径漏改了。
        let (st, d) = try await authed("GET", "\(config.apiBase)/api/recordings/\(id)",
                                        headers: { try self.authHeaders })
        guard let o = try json(st, d, "查会话") as? [String: Any],
              let s = o["session"] as? [String: Any] else { throw DeepBrainError.badResponse("没有 session") }
        return SessionState(status: s["status"] as? String ?? "未知",
                            transcriptId: s["final_transcript_id"] as? String,
                            errorCode: s["last_error_code"] as? String)
    }
}

public extension DeepBrain {
    /// 用邮箱密码登录，换回 refresh token 存进钥匙串。
    ///
    /// 在此之前 Swift 版没有登录能力——靠 Python 版留下的钥匙串条目活着，
    /// 换一台机器直接跑不起来。这是"给别人用"的硬阻塞。
    ///
    /// 密码只用于换 token，不落盘、不打印、不进日志。
    func signIn(email: String, password: String) async throws {
        let (st, data) = try await request(
            "POST", "\(configSupabaseURL)/auth/v1/token?grant_type=password",
            headers: ["apikey": config.supabaseAnonKey],
            json: ["email": email, "password": password])
        guard st < 400,
              let d = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let refresh = d["refresh_token"] as? String,
              let access = d["access_token"] as? String else {
            let body = String(decoding: data, as: UTF8.self)
            // 不要把服务端原文直接抛给用户——里面可能带邮箱
            throw DeepBrainError.http("登录", st,
                body.contains("Invalid login") ? "邮箱或密码不对" : "登录失败")
        }
        TokenStore.set(refresh)
        TokenStore.set(email, "email")
        token = access
        org = try await resolveOrg(access)
    }

    /// 退出：清掉钥匙串里的凭证。
    func signOut() {
        TokenStore.clear()
        token = nil
        org = nil
    }

    static var signedInEmail: String? { TokenStore.get("email") }
    static var hasCredentials: Bool { TokenStore.get() != nil }
}
